### PR TITLE
Gravatar: Interface Layer

### LINKED ISSUE #3492: Allow users to update their Gravatars
#### Description:
It's about the right time to allow the users to update their Gravatar Pictures through WPiOS.

#### Proposed Flow:
1. User taps the current Gravatar to pick or capture a new photo
2. User pans/zooms the photo to select the rectangular area of her liking
3. App uploads and sets the resulting image to Gravatar in a background thread, indicating the operation with some form of progress bar
4. App uses the new Gravatar
5. App dismisses the progress indicator

### PR DESCRIPTION
**Pull request by jleandroperez**
**Note:** This PR is dependant on [this MediaPicker PR](https://github.com/wordpress-mobile/MediaPicker-iOS/pull/107). Let's hold until that is reviewed / merged.

Closes #3492

#### Scenario: Happy Path
1. Log into your WordPress.com account
2. Tap over the `Me` tab
3. Tap over the Gravatar
4. Pick an existing image / take a new one
6. Resize / Crop, as needed
7. Tap over the top right button (`Use`)

As a result, the app should begin uploading the new image immediately, and a spinner should show up on top of the gravatar. Once the spinner is gone, both, **Gravatar.com** and **WPiOS** should display your new image.

#### Scenario: Relaunch
1. Log into your WordPress.com account
2. Tap over the `Me` tab
3. Tap over the Gravatar
4. Pick an existing image / take a new one
6. Resize / Crop, as needed
7. Tap over the top right button (`Use`)
8. Once the upload finishes, please, kill the app, and relaunch.

After relaunch, both, **WPiOS** and **Gravatar.com** should display the same asset.
 
#### Scenario: Multiple Accounts
1. Log into `Account A`, and update its gravatar
2. Log out
3. Log into `Account B`. Verify that `B`'s gravatar is displayed
4. Update `Account B`'s gravatar
5. Log out
6. Log into `Account A`, and verify its gravatar. 

**Please:** Test both iPhone / iPad devices.

Needs review: @astralbodies

### CODE REVIEW

**Comment by astralbodies** [comment]
@jleandroperez I am reviewing this one now!

File: WordPress/Classes/ViewRelated/Media/GravatarOverlayView.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import Foundation
```
Comment (astralbodies) [inline WordPress/Classes/ViewRelated/Media/GravatarOverlayView.swift]: Technically shouldn't this be import UIKit? :smile:

**Comment by astralbodies** [comment]
I'm not sure if this is a result of your PR or if its a long-standing issue. When I log in fresh using a username instead of e-mail address the Me screen looks like:
simulator screen shot apr 5 2016 9 46 26 am

File: WordPress/Classes/ViewRelated/Me/MeHeaderView.m [hunk 2]
```diff
@@ -143,14 +183,51 @@ - (UILabel *)labelForUsername
...
+    if (sender.state == UIGestureRecognizerStateBegan) {
+        [_gravatarImageView depressAnimation];
+        return;
+    }
+    
+    // Touch Up: Normalize the gravatarImageView
+    if (sender.state == UIGestureRecognizerStateEnded) {
+        [_gravatarImageView normalizeAnimation];
+        
+        if (self.onPress) {
```
Comment (astralbodies) [inline WordPress/Classes/ViewRelated/Me/MeHeaderView.m]: Can we make sure that the last touch is still inside the Gravatar view bounds? I'd like to simulate "touch up inside" on this gesture recognizer so you can lift up outside of that view to cancel.

**Comment by astralbodies** [comment]
Logging out and back in with another account shows the previous account's Gravatar until you restart the app.

**Comment by astralbodies** [comment]
I'm setting a Gravatar on an account that I believe doesn't have a Gravatar account set up. The image uploads and gets a success in the console but upon relaunch of the app I do not see a Gravatar.

File: WordPress/Classes/ViewRelated/Me/MeHeaderView.m (hunk 2 above)
Comment (astralbodies) [inline WordPress/Classes/ViewRelated/Me/MeHeaderView.m]: I'd also disable touching this while the image is being uploaded.

**Comment by astralbodies** [comment]
I'm able to tap both the Gravatar image/button and other things on the screen like another tab and the modal UI still presents. Its probably not a huge huge deal but I'm thinking we should not allow other touches.

File: WordPress/Classes/ViewRelated/Media/GravatarOverlayView.swift (hunk 1 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Media/GravatarOverlayView.swift]: The linker isn't complaining, but it won't hurt to add!

File: WordPress/Classes/ViewRelated/Me/MeHeaderView.m (hunk 2 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Me/MeHeaderView.m]: Great points. Updating accordingly, thank you!

**Comment by jleandroperez** [comment]
#### Update:
Just  for self tracking, these are the Done / Pending items. Thanks for the review Aaron!!

- [ ] Caching issues addressed
- [ ] TouchUpInside logic in place
- [ ] Multitouch scenarios prevented
- [ ] [Research first login issue](https://github.com/wordpress-mobile/WordPress-iOS/pull/5058#issuecomment-205840105). Possibly ancient.
- [ ] [Research Gravatar not-set-up issue](https://github.com/wordpress-mobile/WordPress-iOS/pull/5058#issuecomment-205849447)
- [ ] Swiftify Gravatar Helper, and prettify its code

**Comment by jleandroperez** [comment]
@astralbodies ready for another round!. `NSURLCache+Helpers.swift` is meant to fix the universe™.

I'm unable to trigger [this issue](https://github.com/wordpress-mobile/WordPress-iOS/pull/5058#issuecomment-205840105) on my end. Would you please verify if it's still happening on your side as well?. 

That one is possibly ancient, if it's still there, let's address it in another issue (since this PR is already big enough!).

Thanks in advance!

**Comment by astralbodies** [comment]
Reviewing again now!

**Comment by astralbodies** [comment]
This looks super solid to me @jleandroperez. 

:shipit:!!!

**Comment by jleandroperez** [comment]
YAY! Thank you @astralbodies !!!

**Pull request merged**